[BUG][WIP][GENERATOR] Reserved words are handled inconsitently and wrong

Fjolnir-Dvorak opened the pull request:
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

The generator has a feature called isReservedWord which is testing a  word if it is reserved in that language or not. The huge problem is that it ignores case why it is not used by most of the generators, got repaired the wrong way introducing new bugs, and got repaired the correct way, but only on a per language base.

This pull request reimplements the feature isReservedWord. The variable reservedWord is completely removed along with its getter and two new private variiables were added, one list ignoring case and one which is case sensitive. The function isReservedWord will check against those two lists whether the word is blocked or not. To register a word to the blocked list I added the functions `registerReservedWordsCaseSensitive` and `registerReservedWordsCaseInsensitive`. 

For that to work I had to touch each generator and repair the handling of reserved words. That 'caused the java, php and python generator to behave differently than before which must be checked whether it behaves as it should now and did it wrong all the time before, or if I destroyed the code.


The samples will be pushed at last to keep an easy to read an easy to maintain git history

This fix was made because I wanted to fix issue #3428 at first. On further study I noticed the "mess" in the generators why I normalized this feature to make the mentioned bug easy to fig as well as easy to test. The only thing which need to be done to fix that issue is to use the function `registerReservedWordsCaseSensitive` instead of `registerReservedWordsCaseInsensitive`. I did not integrate this change yet to keep the sample changes minimal to make it easier to eliminate bugs before I touch every single sample file of the python generator.

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractAdaCodegen.java:
```diff
@@ -52,7 +66,8 @@ public AbstractAdaCodegen() {
          * Reserved words.  Override this with reserved words specific to your language
          */
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
```
could you please answer this question @stcarrez @micheleISEP?

stcarrez replied:
Ada is case insensitive.

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractPhpCodegen.java:
```diff
@@ -68,7 +79,8 @@ public AbstractPhpCodegen() {
         modelPackage = invokerPackage + "\\" + modelDirName;
 
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
```
in PHP, variable and class names are case insensitive

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppPistacheServerCodegen.java:
```diff
@@ -82,7 +90,8 @@ public CppPistacheServerCodegen() {
 
-        reservedWords = new HashSet<>();
+        // TODO @Fjolnir-Dvorak
+//        reservedWords = new HashSet<>();
```
is this still needed?

Fjolnir-Dvorak replied:
I do not understand the intention of setting reservedWords to an empty HashSet overwriting the parent behavior so I added that just to be save to not accidentally destroy that generator.

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavascriptClientCodegen.java:
```diff
@@ -106,7 +123,8 @@ public JavascriptClientCodegen() {
         // reference: http://www.w3schools.com/js/js_reserved.asp
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(
```
those keywords are case sensitive, but one needs to be added starting with an uppercase as well:
`Boolean`

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavascriptClosureAngularClientCodegen.java:
```diff
@@ -46,7 +61,8 @@ public JavascriptClosureAngularClientCodegen() {
         supportsInheritance = false;
-        setReservedWordsLowerCase(Arrays.asList("abstract",
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(Arrays.asList("abstract",
```
case sensitive

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavascriptFlowtypedClientCodegen.java:
```diff
@@ -43,7 +48,8 @@ public JavascriptFlowtypedClientCodegen() {
 
-        setReservedWordsLowerCase(Arrays.asList(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(Arrays.asList(
```
case sensitive

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/MysqlSchemaCodegen.java:
```diff
@@ -66,7 +78,8 @@ public MysqlSchemaCodegen() {
 
         // https://dev.mysql.com/doc/refman/8.0/en/keywords.html
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
```
case insensitive

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/NodeJSServerCodegen.java:
```diff
@@ -94,7 +108,8 @@ public NodeJSServerCodegen() {
          */
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(
```
case sensitive

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PhpSilexServerCodegen.java:
```diff
@@ -50,7 +59,8 @@ public PhpSilexServerCodegen() {
 
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(
```
case insensitive

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PhpSymfonyServerCodegen.java:
```diff
@@ -103,7 +119,8 @@ public PhpSymfonyServerCodegen() {
 
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(
```
case insensitive

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptFetchClientCodegen.java:
```diff
@@ -254,31 +260,19 @@ private void addOperationObjectResponseInformation(Map<String, Object> operation
-        this.reservedWords.add("VoidApiResponse");
-        this.reservedWords.add("BlobApiResponse");
-        this.reservedWords.add("TextApiResponse");
+        super.registerReservedWordsCaseInsensitive(
```
case sensitive

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptInversifyClientCodegen.java:
```diff
@@ -52,7 +62,8 @@ public TypeScriptInversifyClientCodegen() {
 
-        this.reservedWords.add("map");
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(Collections.singletonList("map"));
```
case sensitive

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptRxjsClientCodegen.java:
```diff
@@ -308,24 +311,29 @@ private void addConditionalImportInformation(Map<String, Object> operations) {
-        this.reservedWords.add("AjaxRequest");
-        this.reservedWords.add("AjaxResponse");
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseSensitive(
```
case sensitive

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

macjohnny commented on modules/openapi-generator/src/main/resources/codegen/generatorClass.mustache:
```diff
@@ -110,8 +110,7 @@ public class {{generatorClass}} extends DefaultCodegen implements CodegenConfig
      * Reserved words.  Override this with reserved words specific to your language
-     */
-    reservedWords = new HashSet<String> (
+     */ super.registerReservedWordsCaseSensitive(
```
add newline

Fjolnir-Dvorak replied:
Resolved in commit [0860bd06](https://github.com/OpenAPITools/openapi-generator/commit/0860bd06dd94abff420fafac93ffb7a12607fef1)

macjohnny commented:
@Fjolnir-Dvorak you could ask the members of the respective technical committees for help: https://github.com/OpenAPITools/openapi-generator#62---openapi-generator-technical-committee

richardwhiuk commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustServerCodegen.java:
```diff
@@ -24,22 +24,40 @@
 import io.swagger.v3.oas.models.parameters.RequestBody;
 import io.swagger.v3.oas.models.servers.Server;
 import org.apache.commons.lang3.StringUtils;
-import org.openapitools.codegen.*;
```
Why have you expanded the * imports?

Fjolnir-Dvorak replied:
I configured the project according to the google_checkstyle.xml file provided in the project. According to line 49 it is forbidden to use star imports:
```xml
        <module name="AvoidStarImport"/>
```

And since I touched it my IDE formatted the changed file the imports so that those a checkstyle conform.

Fjolnir-Dvorak replied:
Well, it was the only thing I configured from checkstyle into my IDE. I have this option per default so I checked the checkstyle rules if I can leave that feature on.

richardwhiuk commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/RustServerCodegen.java:
```diff
@@ -103,7 +121,8 @@ public RustServerCodegen() {
          */
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(
```
They are case-sensitive for Rust.

Fjolnir-Dvorak replied:
Resolved in commit [11ff0928](https://github.com/OpenAPITools/openapi-generator/commit/11ff0928fd16a8a9342c43b65026e1c4dc973d3d)

Fjolnir-Dvorak commented:
I am still trying to understand all new bugs I introduced by fixing this issue. The only language which seems really broken now is Java and a little bit of Python.  Could someone of the java team help me to find the issue?

### Team Java
@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04)


Seems like there are a lot of funny Return classes, but only in python and java.
At the moment I am waiting until I fixed those issues before I will accept the changes / comments above.

```diff
        samples/client/petstore/java/feign/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/feign10x/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/google-api-client/docs/Return.md
        samples/client/petstore/java/google-api-client/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/jersey1/docs/Return.md
        samples/client/petstore/java/jersey1/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/jersey2-java6/docs/Return.md
        samples/client/petstore/java/jersey2-java6/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/jersey2-java8/docs/Return.md
        samples/client/petstore/java/jersey2-java8/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/jersey2/docs/Return.md
        samples/client/petstore/java/jersey2/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/okhttp-gson-parcelableModel/docs/Return.md
        samples/client/petstore/java/okhttp-gson-parcelableModel/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/okhttp-gson/docs/Return.md
        samples/client/petstore/java/okhttp-gson/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/rest-assured/docs/Return.md
        samples/client/petstore/java/rest-assured/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/resteasy/docs/Return.md
        samples/client/petstore/java/resteasy/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/resttemplate-withXml/docs/Return.md
        samples/client/petstore/java/resttemplate-withXml/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/resttemplate/docs/Return.md
        samples/client/petstore/java/resttemplate/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/retrofit/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/retrofit2-play24/docs/Return.md
        samples/client/petstore/java/retrofit2-play24/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/retrofit2-play25/docs/Return.md
        samples/client/petstore/java/retrofit2-play25/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/retrofit2-play26/docs/Return.md
        samples/client/petstore/java/retrofit2-play26/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/retrofit2/docs/Return.md
        samples/client/petstore/java/retrofit2/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/retrofit2rx/docs/Return.md
        samples/client/petstore/java/retrofit2rx/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/retrofit2rx2/docs/Return.md
        samples/client/petstore/java/retrofit2rx2/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/vertx/docs/Return.md
        samples/client/petstore/java/vertx/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/java/webclient/docs/Return.md
        samples/client/petstore/java/webclient/src/test/java/org/openapitools/client/model/ReturnTest.java
        samples/client/petstore/python-asyncio/docs/Return.md
        samples/client/petstore/python-asyncio/petstore_api/models/return.py
        samples/client/petstore/python-asyncio/test/test_return.py
        samples/client/petstore/python-tornado/docs/Return.md
        samples/client/petstore/python-tornado/petstore_api/models/return.py
        samples/client/petstore/python-tornado/test/test_return.py
        samples/client/petstore/python/docs/Return.md
        samples/client/petstore/python/petstore_api/models/return.py
        samples/client/petstore/python/test/test_return.py
        samples/openapi3/client/petstore/python/docs/Return.md
        samples/openapi3/client/petstore/python/petstore_api/models/return.py
        samples/openapi3/client/petstore/python/test/test_return.py
        samples/server/petstore/java-msf4j/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/jaxrs-datelib-j8/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/jaxrs-jersey/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/jaxrs-spec-interface/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/jaxrs-spec/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/jaxrs/jersey1-useTags/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/jaxrs/jersey1/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/jaxrs/jersey2-useTags/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/jaxrs/jersey2/src/gen/java/org/openapitools/model/Return.java
        samples/server/petstore/spring-mvc-j8-async/src/main/java/org/openapitools/model/Return.java
        samples/server/petstore/spring-mvc-j8-localdatetime/src/main/java/org/openapitools/model/Return.java
        samples/server/petstore/spring-mvc/src/main/java/org/openapitools/model/Return.java

```

Fjolnir-Dvorak commented:
Since I am still testing the tests there may come a force push (if possible) to clean up history. The tests are bugged since those do not consider not longer valid files why I tried to delete everything to see exactly which tests would fail to make it easily possible to execute all the tests locally to ensure that the remote tests will work and do what they should do.

richardwhiuk commented:
On the tidy history, note this repo uses squash & merge, so it'll get tidied up before merge into master anyway 😃

Fjolnir-Dvorak commented:
> On the tidy history, note this repo uses squash & merge, so it'll get tidied up before merge into master anyway smiley

@richardwhiuk  Oh sorry, you forked my fork? If you want I can either repair your fork or redo those changes with revert commits. I thought it would be better to force considering the amount of file changes I made.

Fjolnir-Dvorak commented:
Ah, thank you, so it would bnot make a change if I reverted the changes to keep git small and fast. Thank you for the note. I will keep it in my head for next time I decide to split a merge request

Fjolnir-Dvorak commented:
There are still some generators from which I do not know how to configure them:

- [ ] AbstractCppCodegen.java
- [ ] AbstractEiffelCodegen.java
- [ ] AbstractGraphQLCodegen.java
- [ ] AbstractRubyCodegen.java
- [ ] AndroidClientCodegen.java
- [ ] ApexClientCodegen.java
- [ ] BashClientCodegen.java
- [ ] CLibcurlClientCodegen.java
- [ ] CppTizenClientCodegen.java
- [ ] DartClientCodegen.java
- [ ] ElixirClientCodegen.java
- [ ] ElmClientCodegen.java
- [ ] ErlangClientCodegen.java
- [ ] ErlangProperCodegen.java
- [ ] ErlangServerCodegen.java
- [ ] FlashClientCodegen.java
- [ ] GoGinServerCodegen.java
- [ ] GoServerCodegen.java
- [ ] HaskellHttpClientCodegen.java
- [ ] JavascriptFlowtypedClientCodegen.java
- [ ] LuaClientCodegen.java
- [ ] MysqlSchemaCodegen.java
- [ ] NodeJSServerCodegen.java
- [ ] ObjcClientCodegen.java
- [ ] PerlClientCodegen.java
- [ ] PhpSilexServerCodegen.java
- [ ] PowerShellClientCodegen.java
- [ ] RClientCodegen.java
- [ ] ScalaAkkaClientCodegen.java
- [ ] ScalaFinchServerCodegen.java
- [ ] ScalaGatlingCodegen.java
- [ ] ScalaHttpClientCodegen.java
- [ ] ScalaLagomServerCodegen.java
- [ ] ScalatraServerCodegen.java
- [ ] ScalazClientCodegen.java

Should I just ignore those and let them be case insensitive or shall I reference all contributors of those lanugages?

macjohnny commented:
> reference all contributors of those lanugages?

I think this is a good option

daurnimator commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/LuaClientCodegen.java:
```diff
@@ -67,7 +82,8 @@ public LuaClientCodegen() {
         hideGenerationTimestamp = Boolean.TRUE;
 
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
```
case sensitive

eriktim commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ElmClientCodegen.java:
```diff
@@ -75,7 +95,8 @@ public ElmClientCodegen() {
 
-        reservedWords = new HashSet<>(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(
```
for Elm it should be case sensitive

ybelenko commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractPhpCodegen.java:
```diff
@@ -68,7 +79,7 @@ public AbstractPhpCodegen() {
-        setReservedWordsLowerCase(
+        super.registerReservedWordsCaseInsensitive(
                 Arrays.asList(
                         // local variables used in api methods (endpoints)
```
It's absolutely correct, that reserved words("abstract", "and", "array", "as", "break", "callable", "case", "catch", "class") in PHP are case insensitive. However variables aren't. You can check it with simple command:
```bash
$ php -r '$httpBody = "FooBar"; $httpbody = "foobar"; echo $httpBody . PHP_EOL; echo $httpbody;'
// FooBar
// foobar
```
Just keep it in mind, I wouldn't say that it breaks something, but we should be precise in syntax language related topics. Anyway, it's good to warn people not to use `$httpbody` variable when generator contains `$httpBody` to avoid confusion while it's 100% acceptable for PHP parser.

ybelenko commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/FlashClientCodegen.java:
```diff
@@ -81,7 +86,8 @@ public FlashClientCodegen() {
         // from
-        setReservedWordsLowerCase(Arrays.asList("add", "for", "lt", "tellTarget", "and",
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(Arrays.asList("add", "for", "lt", "tellTarget", "and",
```
As far as I remember(I was ActionScript3 developer long time ago and unfortunately don't have Flash Builder right now to check). AS1 was case insensitive, but AS2 and AS3 are case sensitive and highly typed languages. I don't know what version of ActionScript this generator targeted, but assume that AS3.

Fjolnir-Dvorak replied:
Then I will let them to be case insensitive just to be sure, does that sound ok?

beatcracker commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PowerShellClientCodegen.java:
```diff
@@ -88,7 +100,9 @@ public PowerShellClientCodegen() {
         // https://richardspowershellblog.wordpress.com/2009/05/02/powershell-reserved-words/
-        reservedWords = new HashSet<String>(Arrays.asList(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(
```
PowerShell is generally case-insensitive.

robertoaloi commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/ErlangProperCodegen.java:
```diff
@@ -65,7 +77,8 @@ public ErlangProperCodegen() {
         embeddedTemplateDir = templateDir = "erlang-proper";
 
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
```
For Erlang, these are case sensitive. FYI @jfacorro

yissachar commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/DartClientCodegen.java:
```diff
@@ -71,7 +84,8 @@ public DartClientCodegen() {
 
-        setReservedWordsLowerCase(
+        // TODO Are those case sensitive or insensitive
+        super.registerReservedWordsCaseInsensitive(
```
These are case sensitive for Dart

Fjolnir-Dvorak commented:
@spacether could you please look into the failing builds? something with python-experimental is broken now since the reserved word bug is fixed now. Looks like those tests  are based on that broken feature. Can youi fix that or show me where I need to fix that bug, please?

spacether commented:
> @spacether could you please look into the failing builds? something with python-experimental is broken now since the reserved word bug is fixed now. Looks like those tests are based on that broken feature. Can you fix that or show me where I need to fix that bug, please?

Yup, I'll look into this.

spacether commented:
@Fjolnir-Dvorak this error is happening because when we try to get a model's schema here:
https://github.com/Fjolnir-Dvorak/openapi-generator/blob/bug/pythonModelRenamed/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java#L279
ModelUtils.getSchema(this.openAPI, cm.name) returns null as the model's schema.
Why is the model's schema null? It should exist.

Fjolnir-Dvorak commented:
Thank you very much. Will look into it

Fjolnir-Dvorak commented:
That is a lovely problem. The name does not get set, not even to null.
since CodegenModel does not even have a constructor it will take some time to find a setter which does not exist.

Fjolnir-Dvorak commented:
@spacether I think there different sanitizers used here. I added some debug log and got boiled down to following issue:

```
[main] ERROR o.o.codegen.utils.ModelUtils - no Schema _Return found in Schematas: [Order, Category, User, Tag, Pet, ApiResponse, $special[model.name], Return, Name, 200_response, ClassModel, Dog, Cat, Animal, AnimalFarm, format_test, EnumClass, Enum_Test, AdditionalPropertiesClass, AdditionalPropertiesString, AdditionalPropertiesInteger, AdditionalPropertiesNumber, AdditionalPropertiesBoolean, AdditionalPropertiesArray, AdditionalPropertiesObject, AdditionalPropertiesAnyType, MixedPropertiesAndAdditionalPropertiesClass, List, Client, ReadOnlyFirst, hasOnlyReadOnly, Capitalization, MapTest, ArrayTest, NumberOnly, ArrayOfNumberOnly, ArrayOfArrayOfNumberOnly, EnumArrays, OuterEnum, OuterComposite, OuterNumber, OuterString, OuterBoolean, StringBooleanMap, FileSchemaTestClass, File, TypeHolderDefault, TypeHolderExample, XmlItem, Dog_allOf, Cat_allOf]
[main] ERROR o.o.codegen.DefaultCodegen - Undefined property/schema for `value`. Default to type:string.
[main] ERROR o.o.codegen.DefaultCodegen - Undefined property/schema for `value`. Default to type:string.
Exception in thread "main" java.lang.NullPointerException
        at org.openapitools.codegen.languages.PythonClientExperimentalCodegen.fromProperty(PythonClientExperimentalCodegen.java:304)
        at org.openapitools.codegen.languages.PythonClientExperimentalCodegen.postProcessAllModels(PythonClientExperimentalCodegen.java:273)
        at org.openapitools.codegen.DefaultGenerator.generateModels(DefaultGenerator.java:480)
        at org.openapitools.codegen.DefaultGenerator.generate(DefaultGenerator.java:923)
        at org.openapitools.codegen.cmd.Generate.run(Generate.java:416)
        at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:61)

```

You are trying to get a Schema for "_Return" but since as far as I know Python is case sensitive so the model is named "Return".

spacether commented:
@Fjolnir-Dvorak where is model.name being set? Should it be mutated from what is in the spec?
If yes how do we ensure that modelutils.getschema with the new name will still work?

Fjolnir-Dvorak commented:
Model names are sanitized and checked and escaped int the function https://github.com/Fjolnir-Dvorak/openapi-generator/blob/0925bddabcbc2ae7e3f731ffdabdf358b4f9051a/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientCodegen.java#L531 if they contain a reserved word, that's why xou get the input "_Return". I sadly do not know yet how that modelutils.geschema stuff works but it looks like that this stuff is not escaped and handled properly, so maybe there must be an alterantive which is working?

spacether commented:
How about adding a model property schemaName which stores the unmodified model name? That will preserve our ability to get the model's original schema. Or maybe we can convert the model's json back into a schema object.

spacether commented:
Did you intend to link to the function toModelFilename or did you mean toModelName?

Fjolnir-Dvorak commented:
I think that that link is the one responsible for that since toModelName is not adding an underscore before the name. can be that I am wrong with the specific link, but hte point is that it looks like your input is already sanitized and the schemas not

Fjolnir-Dvorak commented:
> How about adding a model property schemaName which stores the unmodified model name? That will preserve our ability to get the model's original schema. Or maybe we can convert the model's json back into a schema object.

would be an idea / sounds like a way to solve this issue. But since I don't know anything about that part yet I would ask a maintainer how to resolve this.

spacether commented:
@Fjolnir-Dvorak if we were invoking toModelName, then the result of "_Return" would not be set. Something else is happening here to incorrectly set model.name
Per the master branch:
https://github.com/OpenAPITools/openapi-generator/blob/de162f7f34e6ac2bf91b289fd7c6ae9f82e03f4b/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java#L468
toModelName is looks like it is used to set model.className.
The python generator uses className in it's model template:
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/model.mustache#L15
Based on this I don't think that we should be changing the value of model.name, we should be updating model.className with the sanitized name.

spacether commented:
This is happening because we are invoking escapeReservedWord for the model named Return https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L1756
It then sets the name to "_Return"
master branch testing confirms that previously we were going down the code path here:
https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L1758
And setting model name to "Return"
The name value is used to set classname here: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L1763
How about adding the schemaName property to model?
That way I can use it to get the schema. Otherwise we lose the info what was this model originally named.

PythonCodegenExperimental needs this because for modes whose type != object, if they have validations, they cannot be stored in the model object. To preserve them, I modify the model, adding a value parameter which includes the validations.
So string model
```
DashName
   type: string
   pattern: '^\d{3}-\d{2}-\d{4}$'
```
becomes
```
Dashname
   type: object
   properties: value
      type: string
      pattern: '^\d{3}-\d{2}-\d{4}$'
```
To do this, we use the model's schema to make the property value.
If we do not do this, then the DashName model would omit validation checking when it is generated.

Could you add a model.schemaName property, then I can use it to get the model's schema?

Fjolnir-Dvorak commented:
I will look into it this weekend. Thank you for looking where the problem was.

spacether commented:
Welcome. Investigation into our codebase and swagger-codegen has yielded no examples where we go from modelJson back to a Schema object:
- https://github.com/OpenAPITools/openapi-generator/search?q=.modelJson&unscoped_q=.modelJson
- https://github.com/swagger-api/swagger-codegen/search?q=.modelJson&unscoped_q=.modelJson
converting modelJson back to Schema may be possible using some helper functions from other java imported libraries but it is not looking simple at this point.

spacether commented:
@Fjolnir-Dvorak  I have a fix that doesn't require adding the model.schemaName
Here: https://github.com/Fjolnir-Dvorak/openapi-generator/blob/0925bddabcbc2ae7e3f731ffdabdf358b4f9051a/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java#L588
Change it to:
```
        CodegenModel result = super.fromModel(name, schema);
        // hard code the model name to the original name so we can later use it to extract the model's schema
        // in postProcessAllModels
        result.name = name;

```

ybelenko commented:
@spacether There is no progress on this from **Nov 21, 2019** and many many merge conflicts. Should we keep this PR open?

Fjolnir-Dvorak commented:
I will work on it again, soon. Just got demotivated/annoyed by the is-mergeable check and that all buildservers exit early instead of running through all tests.

ybelenko commented:
@Fjolnir-Dvorak almost 2 years passed, should we close it now? Maybe we can use it as reference in next PR.

Fjolnir-Dvorak commented:
I could make it work again. The problem is still the same, is it? Since it was a fix to the core module and a lot of contributors hacked their generator to somehow work or not those are broken with this pull request. I do not know the current state if the generators are decoupled now but if not how should I make this work if there is no interest in all the language-maintainers to fix their generator or to help resolve their code? There were a lot of unanswered design decisions :/

gmkumar2005 commented:
The ScalaLagomServerCodegen.java is out dated. It is not in sync with Lagom framework. Please drop this feature. The Lagom framework itself  is reaching EOL. Hence there would be no users for  ScalaLagomServerCodegen.

ybelenko commented:
> I could make it work again. The problem is still the same, is it? Since it was a fix to the core module and a lot of contributors hacked their generator to somehow work or not those are broken with this pull request. I do not know the current state if the generators are decoupled now but if not how should I make this work if there is no interest in all the language-maintainers to fix their generator or to help resolve their code? There were a lot of unanswered design decisions :/

I share your disappointment. Can we split your enhancement into and apply it partially? I can help with PHP based codegens, mysql and javascript. Might be with shell/bash.

I agree that silent community is the biggest problem of this repo right now. I would allow codegen authors and top contributors to review and merge their own PRs, since it takes months some times to get review of a few code lines. It feels weird, that I can merge somebody else PR since I'm member of technical committee, but I cannot merge my own one. I got the point of the second look, but we have no active reviewers.

Fjolnir-Dvorak commented:
I will look into it, but I think it is difficult. It is a correction of a function/feature which was heavily misused or simply did not work. So fixing it forcefully changes all generators and destroys backwards-compability in a tool where it does not matter in my opinion (since you can fix the version you are using for your generated code). And if we would break and remove no longer or not actively maintained languages mayhaps we would have more people contributing to patch back in all those broken languages to make them work again. Python client for example destroyed their whole generator for me since they removed all typing and type hints and field hints since they went all meta. Well, their generator looks cool, but the code is unusable if you like to work with autocompletion, there is no more autocompletion. Mayhaps that was added in 5.x but it does not matter since it is a generator. If something breaks there is no disadvantage to revert the version of the generator and stick to the old one, mayhaps create your own fork for your company which I did for every language we are using since it is way easier to fix bugs if you do it for yourself and if you have a very long breath and a lot of free time, maybe you can someday contribute those fixes back into upstream. Well, enough about that, sorry :D

I will look into it if it is somehow possible, mayhaps by duplicating code and creating bloatware with deprecated and fixed code if that would be an option. Mayhaps we could generate some warnings and a lot of warnings for generators which are using a lot of deprecated and broken features. That would be an idea, even if it is ugly as something which is ugly. Don't think hell is ugly, at least not in my D&D Setting.

Sorry for all that negativity. OpenApi is one of my favourite projects and thank you a lot for writing that you share the disappointment, but it was or is just impossible to really do and fix something in this project with the current workflow and i do not think that I am the only one who is workflowed away from helping making the generator to an even more awesome project.

spacether commented:
Sorry to hear that the python generator is not meeting your needs if you want it to have type hints how about submitting an issue and PR that adds them on fields? Meta restructuring was done to consolidate where payloads were stored. Also, python-experimental has type hints and may meet your needs, have you tried it?

Fjolnir-Dvorak commented:
@spacether oh sorry, everything is fine with the generator itself, it just went into a direction which I for myself do not like but I am sure it has its purposes. It more of an example of a generator changing quite a bit regarding backwards compability which felt like a strong argument in the past. The python generator is changing which is a good thing. I have my own python generator fork to keep compability for regenerating clients,
